[NHUB-530] Implement async users management

eos87 opened the pull request:
### Purpose
Implement Users Management section in new Async Framework

### What's changed
- Create new module, resource model, service and helpers using async framework 
- Update references and adjusting parts of the app to use the new toys
- Fixed and revisited many tests that got broken because of the adjustments 

### What's pending
- Remove some references to `get_resource_service("users")` which were left there because of the time already used in this tasks
~- Fix a couple of tests that are failing also left out because of the time (behave and e2e)~

Thanks for checking!

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] This pull request is not adding new forms that use redux
- [ ] This pull request is adding missing TypeScript types to modified code segments where it's easy to do so with confidence
- [ ] This pull request is replacing `lodash.get` with optional chaining for modified code segments

eos87 committed:
Add resource, base service and module with new framework

NHUB-530

eos87 committed:
Implement async users' helpers

NHUB-530

eos87 committed:
Implement `/users/search` async view

NHUB-530

eos87 committed:
Complete implementation of `/users/search`

NHUB-530

eos87 committed:
Implement async `/users/edit` view

NHUB-530

eos87 committed:
Implement async `/users/create`

NHUB-530

eos87 committed:
Implemente async `/users/create`

NHUB-530

eos87 committed:
Implement the rest of `users.views` to async

NHUB-530

eos87 committed:
Remove not needed imports

NHUB-530

MarkLark86 commented on newsroom/companies/__init__.py:
```diff
@@ -1,39 +1,32 @@
 from flask_babel import lazy_gettext
 
-import superdesk
+from newsroom.companies.companies_async.types import CompanyResource
```
`CompanyResource` is exposed by the top level module, so you can import by using:

```python
from newsroom.companies import CompanyResource
```

This is a good example of a module having a defined interface of what it exposes. It might be good if we try and follow that with the modules, so we're not importing from too deep within the modules. `from newsroom.<x> import <y>`.

eos87 replied:
yeah, I think this was just auto import from the editor. I'll fix it

MarkLark86 commented on newsroom/companies/__init__.py:
```diff
@@ -1,39 +1,32 @@
-    }
-    if company and company.get("sections"):
-        rv["userSections"] = [s for s in app.sections if company.get("sections").get(s["_id"])]
+    if not company or user.user_type == UserRole.ADMINISTRATOR.value:
```
When checking against an enum, there is no longer the need to use `.value` from the enum attribute.

i.e.
```console
>>> from newsroom.user_roles import UserRole
>>> from pydantic import BaseModel
>>> 
>>> class User(BaseModel):
...     name: str
...     role: UserRole
... 
>>> user = User(name="Mark", role=UserRole.PUBLIC)
>>> user.role
<UserRole.PUBLIC: 'public'>
>>> user.role == UserRole.PUBLIC
True
>>> user.role == UserRole.INTERNAL
False
```

eos87 committed:
Fix flake8 complain

eos87 committed:
Merge remote-tracking branch 'origin/async' into hg/NHUB-530-users-management-async

eos87 committed:
Remove duplicated imports after merge

NHUB-530

eos87 committed:
Fix mypy complains about app.cache missing

NHUB-530

eos87 committed:
Add missing await after merge

NHUB-530

eos87 committed:
Complete pending methods and fix tests

NHUB-530

eos87 committed:
Fix auth and agenda tests

NHUB-530

eos87 committed:
Fix more tests

NHUB-530

MarkLark86 commented on newsroom/users/service.py:
```diff
@@ -0,0 +1,181 @@
+class UsersService(NewshubAsyncResourceService[UserResourceModel]):
+    resource_name = "users"
+
+    def get_model_instance_from_dict(self, data: Dict[str, Any]) -> ResourceModelType:
```
Rather than supported invalid data, what if we have a data update script that fixes the data. That way we don't need this custom code to support invalid data in the DB

eos87 replied:
Yeah, that makes sense. Do you mean like those from this command `python manage.py data:generate_update`? If so, yes we should do it. Although I'm not sure if these commands are working at the moment. Perhaps we should have this in a separate small ticket (also to remove this workaround). Thoughts?

MarkLark86 replied:
Yes that's the one. Those commands should work, and seem to work on my local machine (although we're missing `manage.py` in newsroom-core repo, so I generally run it on a `newsroom-<name>-app` repo and copy the file over.

If you can, have a quick go at it, otherwise I will create a ticket to cover this (and test the commands).

eos87 replied:
Fixed in commits: https://github.com/superdesk/newsroom-core/pull/1036/commits/d437a72dc5ab0378ecb89db8e61a853c2b947dbb https://github.com/superdesk/newsroom-core/pull/1036/commits/c54ddd987aba4988563035cb8ba016218087a64b

You would need to run this below to make it work properly. 
```python manage.py data_upgrade -i 00016_20240820-184145_users```

eos87 committed:
Try to fix broken test without success

NHUB-530

eos87 committed:
Fix broken tests

NHUB-530

eos87 committed:
Clean up and small adjustment

MarkLark86 commented on newsroom/companies/views.py:
```diff
@@ -176,15 +175,16 @@ async def edit_company(args: CompanyItemArgs, params: None, request: Request) ->
     try:
-        # TODO-ASYNC: Convert to users async service
-        get_resource_service("users").delete_action(lookup={"company": args.company_id})
+        await UsersService().delete_many(lookup={"company": str(args.company_id)})
```
It's probably due to the `Store ObjectId in MongoDB from async resource services` bug, but should this be an ObjectId when looking up which users to delete and not a string?

eos87 replied:
yeah, I'll adjust it now that you prepared the fix.

MarkLark86 commented on newsroom/companies/views.py:
```diff
@@ -199,14 +199,22 @@ async def delete_company(args: CompanyItemArgs, params: None, request: Request)
+    from newsroom.users.service import UsersService
+
+    users_list = []
+    company_users = await UsersService().search(lookup={"company": str(args.company_id)})
```
Same goes here

MarkLark86 commented on newsroom/companies/views.py:
```diff
@@ -222,9 +230,10 @@ async def approve_company(args: CompanyItemArgs, params: None, request: Request)
-    users_service = get_resource_service("users")
-    for user in users_service.get(req=None, lookup={"company": original.id, "is_approved": {"$ne": True}}):
+    users_service = UsersService()
+    company_users = await users_service.search(lookup={"company": str(original.id)})
```
And here. Thinking because we have this in a few locations, it might be good to have a common/util function that provides the list of users for a company

eos87 replied:
yep, agree. I'll a separate method to avoid repeating.

eos87 replied:
done.

eos87 committed:
Add `get_users_by_company` helper function

NHUB-530

eos87 committed:
Add data migration to remove `id` and `signup_details` from users

NHUB-530

eos87 committed:
Remove no longer needed method

In order to get the new `UsersService` to work, running next command is needed

`python manage.py data_upgrade -i 00016_20240820-184145_users`

NHUB-530

eos87 committed:
Fix linter

NHUB-530

MarkLark86 reviewed (approved):
Looks like tests are failing.

Behave tests looks like it's trying to add users from Superdesk, as Pydantic validation is complaining about `needs_activation` field which is from Superdesk and not Newshub.

Not sure what's causing the E2E tests to fail

MarkLark86 commented on newsroom/users/views.py:
```diff
@@ -56,196 +75,241 @@ def get_settings_data():
+                add_token_data(new_user)
 
-        ids = get_resource_service("users").post([new_user])
+        ids = await UsersService().create([new_user])
```
Sorry, I missed this one. Would need to wrap `service.create` and `service.update` in a try...except to catch and return validation errors to the browser.

See an example [here](https://github.com/superdesk/newsroom-core/blob/async/newsroom/companies/views.py#L84)

eos87 replied:
done.

eos87 committed:
Add custom behave step to avoid superdesk's users attrs

NHUB-530

eos87 committed:
Fix broken tests

Instead of continue relying on dict we're better using model attributers directly which also offer proper typings

NHUB-530

eos87 committed:
Wrap validation errors and return them to user

NHUB-530

eos87 committed:
Fix e2e tests

NHUB-530
